src/Dataflow/Core/Node.hpp

```diff
@@ -0,0 +1,546 @@
+    using PortBaseCollection = PortCollection<PortBasePtr>;
+
+    using PortBaseInPtr        = PortPtr<PortBaseIn>;
+    using PortBaseInRawPtr     = PortRawPtr<PortBaseIn>;
+    using PortBaseInCollection = PortCollection<PortBaseInPtr>;
+
+    using PortBaseOutPtr        = PortPtr<PortBaseOut>;
+    using PortBaseOutRawPtr     = PortRawPtr<PortBaseOut>;
+    using PortBaseOutCollection = PortCollection<PortBaseOutPtr>;
+
+    template <typename Port>
+    using IndexAndPort = std::pair<PortIndex, Port>;
+
+    template <typename T>
+    using ParamHandle = Ra::Core::VariableSet::VariableHandle<T>;
+
+    /// \name Constructors
+    /// @{
+    /// \brief delete default constructors.
+    /// \see https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rc-copy-virtual
+    Node()                         = delete;
+    Node( const Node& )            = delete;
+    Node& operator=( const Node& ) = delete;
+    /// @}
+
+    /// \brief make Node a base abstract class
+    virtual ~Node() = default;
+
+    /// \brief Two nodes are considered equal if there model and instance names are the same.
+    bool operator==( const Node& node );
+
+    /// \name Function execution control
+    /// @{
+    /**
+     * \brief Initializes the node content.
+     *
+     * The init() function should be called once at the beginning of the lifetime of the node by
+     * the owner of the node (the graph which contains the node).
+     * Its goal is to initialize the node's internal data if any.
+     * The default do nothing.
+     */
+    virtual void init();
+
+    /**
+     *\brief Compile the node to check its validity.
+     *
+     * Only nodes defining a full computation graph will need to override this method.
+     * The base version do nothing.
+     * \return the compilation status
+     */
+    virtual bool compile();
+
+    /**
+     * \brief Executes the node.
+     *
+     * Execute the node function on the input ports (to be fetched) and write the results to the
+     * output ports.
+     * \return the execution status.
+     */
+    virtual bool execute() = 0;
+
+    /**
+     * \brief Delete the node's content.
+     *
+     * The destroy() function is called once at the end of the lifetime of the node.
+     * Its goal is to free the internal data that have been allocated.
+     */
+    virtual void destroy();
+    /// @}
+
+    /// \name Control the interfaces of the nodes (inputs, outputs, internal data, ...)
+    /// @{
+
+    /**
+     * \brief Get a port by its name
+     *
+     * \param type either "in" or "out", the directional type of the port
+     * \param name
+     * \return the index to access the port and a raw ptr to the port.
+     */
+    auto port_by_name( const std::string& type, const std::string& name ) const
+        -> IndexAndPort<PortBaseRawPtr>;
+
+    /// Convenience alias to port_by_name("in", name)
+    IndexAndPort<PortBaseInRawPtr> input_by_name( const std::string& name ) const;
```

Comment:
Should we choose only one type of function declaration, i.e `auto f(...)-> returnType;` or `returnType f(...);`. I vote the first but agree with the other also.

Reply:
I think we can use auto for complex/templated/nested return value. and keep plain return for simple/std types. I'll change it.